### PR TITLE
Enable region based selection

### PR DESCRIPTION
**Pull request by hamza-vd**
See #98

### CODE REVIEW

File: opensrp-eusm/src/main/java/org/smartregister/eusm/repository/AppLocationRepository.java [hunk 1]
```diff
@@ -71,7 +71,7 @@ public Set<Location> getLocationByNameAndGeoLevel(Set<String> names, String leve
         if (names == null || names.isEmpty())
             return locations;
         try (Cursor cursor = getReadableDatabase().rawQuery("SELECT * FROM " + getLocationTableName() +
-                " WHERE " + StringUtils.repeat(NAME + " =?", " OR ", names.size()) + " AND " + GEOGRAPHICAL_LEVEL + "=?", ArrayUtils.add(names.toArray(new String[0]), level))) {
+                " WHERE " + StringUtils.repeat("(" + NAME + " =? AND " + GEOGRAPHICAL_LEVEL + " = " + level + ")", " OR ", names.size()), names.toArray(new String[0]))) {
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/repository/AppLocationRepository.java:74]:
```suggestion
                " WHERE (" + StringUtils.repeat(NAME + " =?", " OR ", names.size()) + ") AND " + GEOGRAPHICAL_LEVEL + "=?", ArrayUtils.add(names.toArray(new String[0]), level))) {
```
The previous query can still work by surrounding it with parenthesis.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/view/EusmTreeViewDialog.java [hunk 2]
```diff
@@ -27,19 +29,31 @@ public EusmTreeViewDialog(Context context, int theme, JSONArray structure,
 
     @Override
     public void init(JSONArray nodes, ArrayList<String> defaultValue, ArrayList<String> value, boolean isSelectionMode) throws JSONException {
-        this.operationalAreas = PreferencesUtil.getInstance().getCurrentOperationalAreas();
+        try {
+            // Get selected regions from districts
+            this.operationalRegions = populateOperationalRegions(nodes);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/view/EusmTreeViewDialog.java:36]: Use timber here.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/view/EusmTreeViewDialog.java [hunk 3]
```diff
@@ -27,19 +29,31 @@ public EusmTreeViewDialog(Context context, int theme, JSONArray structure,
...
     @Override
     public TreeNode.BaseNodeViewHolder<String> getNodeViewHolder(Context context, JSONObject structure) {
         return new TreeSelectionItemViewHolder(context, structure.optString(KEY_LEVEL, ""));
     }
 
     @Override
     public void updateTreeNode(int level, TreeNode curNode, TreeNode parentNode) {
-        curNode.setSelectable(level == 3);
-        boolean isSelected = operationalAreas.contains(curNode.getValue().toString());
+        curNode.setSelectable(level == 2);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/view/EusmTreeViewDialog.java:55]: replace 2 with `LocationGeographicLevel` interface value.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/view/NavigationDrawerView.java [hunk 4]
```diff
@@ -223,4 +225,27 @@ public void toggleProgressBarView(boolean syncing) {
         View flexBox = this.activity.getActivity().findViewById(R.id.flex_box);
         flexBox.setVisibility((syncing && NetworkUtils.isNetworkAvailable()) ? View.GONE : View.VISIBLE);
     }
+
+    @Override
+    public void setOperationalArea(String operationalArea) {
+        try {
+            super.setOperationalArea(getRegionsFromDistricts(operationalArea));
+        } catch (Exception e) {
+            Timber.e(e);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/view/NavigationDrawerView.java:234]: Use timber here.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/view/NavigationDrawerView.java [hunk 5]
```diff
@@ -223,4 +225,27 @@ public void toggleProgressBarView(boolean syncing) {
...
+        }
+    }
+
+    private String getRegionsFromDistricts(String districts) throws Exception{
+        if (districts == null) {
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/view/NavigationDrawerView.java:239]: Check for empty string too.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java [hunk 6]
```diff
@@ -249,4 +251,27 @@ public static void updateLocationCoordinatesFromForm(StructureDetail structureDe
             structureRepository.addOrUpdate(location);
         }
     }
+
+    public static ArrayList<String> getRegionsForDistricts(JSONArray locationHierarchy, Set<String> districts, boolean hideDistricts) throws Exception {
+        ArrayList<String> opRegions = new ArrayList<>();
+        JSONArray regions = locationHierarchy.getJSONObject(0).getJSONArray(AppConstants.JsonForm.NODES);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java:257]: Check for null on `.getJSONObject(0)` and `regions`

File: opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java [hunk 7]
```diff
@@ -249,4 +251,27 @@ public static void updateLocationCoordinatesFromForm(StructureDetail structureDe
...
+        for (int i = 0; i < regions.length(); i++) {
+            JSONObject location = regions.getJSONObject(i);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java:259]: Check for null on `location`.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java [hunk 8]
```diff
@@ -249,4 +251,27 @@ public static void updateLocationCoordinatesFromForm(StructureDetail structureDe
...
+            String locName = location.getString(TaskingConstants.CONFIGURATION.KEY);
+            if (location.has(AppConstants.JsonForm.NODES)) {
+                JSONArray childNodes = location.getJSONArray(AppConstants.JsonForm.NODES);
+                for (int j = 0; j < childNodes.length(); j++) {
+                    if (districts.contains(childNodes.getJSONObject(j).getString(TaskingConstants.CONFIGURATION.KEY))) {
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java:264]: Check for null `childNodes.getJSONObject(j).`

File: opensrp-eusm/src/main/java/org/smartregister/eusm/view/EusmTreeViewDialog.java (hunk 3 above)
Comment (bennsimon) [reply opensrp-eusm/src/main/java/org/smartregister/eusm/view/EusmTreeViewDialog.java:55]: Add comment above the method too.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java [hunk 9]
```diff
@@ -81,61 +86,67 @@ public void run() {
...
     /**
      * Filter out district not in the plan
      *
-     * @param filteredIn
+     * @param jurisdictionList
      * @param entireTree
      * @return
      */
-    private List<FormLocation> filterLocations(List<String> filteredIn, List<FormLocation> entireTree) {
+    private List<FormLocation> filterLocations(List<String> jurisdictionList, List<FormLocation> entireTree) {
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java:139]: Update the comment above the method is needed.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java [hunk 10]
```diff
@@ -81,61 +86,67 @@ public void run() {
         }
     }
 
+    @Override
+    public void onOperationalAreaSelectorClicked(ArrayList<String> names) {
+        // Extract districts from selected regions
+        ArrayList<String> districts = new ArrayList<>();
+        try {
+            JSONArray locationHierarchy = new JSONArray(extractLocationHierarchy().first);
+            JSONArray regions = locationHierarchy.getJSONObject(0).getJSONArray(AppConstants.JsonForm.NODES);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java:95]: Check for null `getJSONObject(0)` and finally `regions`.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java [hunk 11]
```diff
@@ -81,61 +86,67 @@ public void run() {
...
+            for (int i = 0; i < regions.length(); i++) {
+                JSONObject location = regions.getJSONObject(i);
+                String locName = location.getString(TaskingConstants.CONFIGURATION.KEY);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java:98]: Check for null  on `locations`.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java [hunk 12]
```diff
@@ -81,61 +86,67 @@ public void run() {
...
+                if (names.contains(locName)) {
+                    names.remove(locName);
+                    if (location.has(AppConstants.JsonForm.NODES)) {
+                        JSONArray nodes = location.getJSONArray(AppConstants.JsonForm.NODES);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java:102]: Check for null on `nodes`.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java [hunk 13]
```diff
@@ -249,4 +251,27 @@ public static void updateLocationCoordinatesFromForm(StructureDetail structureDe
...
     }
+
+    public static ArrayList<String> getRegionsForDistricts(JSONArray locationHierarchy, Set<String> districts, boolean hideDistricts) throws Exception {
+        ArrayList<String> opRegions = new ArrayList<>();
+        JSONArray regions = locationHierarchy.getJSONObject(0).getJSONArray(AppConstants.JsonForm.NODES);
+        for (int i = 0; i < regions.length(); i++) {
+            JSONObject location = regions.getJSONObject(i);
+            String locName = location.getString(TaskingConstants.CONFIGURATION.KEY);
+            if (location.has(AppConstants.JsonForm.NODES)) {
+                JSONArray childNodes = location.getJSONArray(AppConstants.JsonForm.NODES);
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java:262]: Check for null on `childNodes`.

File: opensrp-eusm/src/main/java/org/smartregister/eusm/repository/AppLocationRepository.java (hunk 1 above)
Comment (bennsimon) [reply opensrp-eusm/src/main/java/org/smartregister/eusm/repository/AppLocationRepository.java:74]: The above should achieve the existing parameter binding.

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java (hunk 12 above)
Comment (hamza-vd) [reply opensrp-eusm/src/main/java/org/smartregister/eusm/presenter/EusmBaseDrawerPresenter.java:102]: `nodes` will not be null as we already have a check at line `101` which make sure that the current location has nodes

File: opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java (hunk 13 above)
Comment (hamza-vd) [reply opensrp-eusm/src/main/java/org/smartregister/eusm/util/AppUtils.java:262]: `childNodes` will not be null as we already have a check in place at line `261` which will make sure that the current location has nodes

File: opensrp-eusm/src/main/java/org/smartregister/eusm/view/NavigationDrawerView.java [hunk 14]
```diff
@@ -236,7 +236,7 @@ public void setOperationalArea(String operationalArea) {
     }
 
     private String getRegionsFromDistricts(String districts) throws Exception{
-        if (districts == null) {
+        if (StringUtils.isEmpty(districts)) {
```
Comment (bennsimon) [inline opensrp-eusm/src/main/java/org/smartregister/eusm/view/NavigationDrawerView.java:239]: Use `isBlank` instead it checks for `null`, `empty` and string with spaces.

**Review by bennsimon — CHANGES_REQUESTED** [review]

**Review by bennsimon — APPROVED** [review]

**Pull request merged**